### ISSUE TITLE
02-linear-regression-reg-medical-insurance-cost.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/proyecto-machine-learning-gregoriant#3]
Hola @gregangel1550-del  el EDA excelente pero....

### **1. CRÍTICO - fit_transform en X_test (Data Leakage)**

**Celda de scaling para test:**
```python
min_max_features_test = min_max_scaler.fit_transform(X_test[num_variables])  # ❌❌ fit_transform

df_min_max_test = pd.DataFrame(min_max_features_test,
                          index=X_test.index,
                          columns=num_variables)
```

**PROBLEMA GRAVE:** Aplicas `fit_transform()` en X_test, lo que significa que calcula nuevas estadísticas (min, max) basadas en los datos de test. Esto causa **data leakage** porque:
- El modelo "ve" información de test durante el entrenamiento
- Las estadísticas de normalización son diferentes entre train y test
- Los datos no están en la misma escala, invalidando las predicciones

**DEBERÍA SER:**
```python
min_max_features_test = min_max_scaler.transform(X_test[num_variables])  # Solo transform
```

---

### **2. GRAVE - No incluyes variable "region" en el encoding**

**Variables categóricas definidas:**
```python
cat_variables = ['sex', 'smoker']  # Falta 'region'
```

**PROBLEMA:** La variable `region` tiene 4 categorías (southwest, southeast, northwest, northeast) y es categórica, pero NO se incluye en el OneHotEncoder. Esto significa que:
- La variable `region` nunca se procesa
- No se incluye en el modelo final
- Se pierde información potencialmente valiosa

Analizas `region` en el EDA pero luego la olvidas en el encoding.

---

### **3. GRAVE - No concatenas features procesadas**

Después de escalar las numéricas y codificar las categóricas:
- `df_min_max_train` tiene las numéricas escaladas: age, bmi, children
- `X_train_cat_ohe` tiene las categóricas codificadas: sex_female, sex_male, smoker_no, smoker_yes

**Pero NO las concatenas**. En su lugar, usas SelectKBest solo sobre las categóricas:

```python
selection_model.fit(X_train_cat_ohe, y_train)  # Solo usas categóricas
X_train_sel = selection_model.transform(X_train_cat_ohe)
X_test_sel = selection_model.transform(X_test_cat_ohe)
```

**PROBLEMA GRAVE:** 
- El modelo final solo usa las variables categóricas codificadas (sex, smoker)
- Las variables numéricas escaladas (age, bmi, children) nunca se usan en el modelo
- Esto explica por qué SelectKBest tiene k=5: solo hay 4 columnas one-hot (sex_female, sex_male, smoker_no, smoker_yes)

**DEBERÍA SER:**
```python
# Concatenar numéricas escaladas + categóricas codificadas
X_train_final = pd.concat([df_min_max_train, X_train_cat_ohe], axis=1)
X_test_final = pd.concat([df_min_max_test, X_test_cat_ohe], axis=1)

# Luego aplicar SelectKBest si es necesario
```

---

### **4. SelectKBest con k=5 sobre solo 4 columnas**

```python
selection_model = SelectKBest(f_classif, k=5)
selection_model.fit(X_train_cat_ohe, y_train)
```

**PROBLEMA:** 
- `X_train_cat_ohe` tiene 4 columnas (sex_female, sex_male, smoker_no, smoker_yes)
- SelectKBest pide k=5 (seleccionar 5 features)
- No se puede seleccionar 5 de 4, así que devuelve las 4

Esto es técnicamente un error menor porque SelectKBest simplemente devuelve todas las columnas, pero indica confusión sobre qué datos se están usando.

---

### **1. SelectKBest debería aplicarse DESPUÉS de concatenar**

SelectKBest está pensado para seleccionar las mejores k features de TODAS las variables disponibles. Al aplicarlo solo sobre las categóricas, pierde su propósito.

---

## RESUMEN:

**Errores críticos que invalidan el modelo:**
1. **fit_transform en X_test** - Data leakage que invalida las métricas
2. **No concatenas features procesadas** - El modelo solo usa sex y smoker
3. **Ignoras variables numéricas escaladas** - age, bmi, children no se usan
4. **Ignoras variable region** - Pierde información categórica importante
5. **SelectKBest mal aplicado** - Solo sobre categóricas con k incorrecto

**Impacto en resultados:**
- El R² reportado está completamente inflado por el data leakage en scaling
- El modelo está severamente limitado al usar solo 2 variables (sex, smoker) cuando debería usar 6 (age, bmi, children, sex, smoker, region)
- Las predicciones son inválidas porque train y test están en escalas diferentes

**Veredicto:** A pesar de un EDA excelente y buenas intenciones metodológicas, cometes tres errores críticos que invalidan completamente el modelo: data leakage en el scaling de test, no concatenar las features procesadas, y consecuentemente entrenar un modelo que solo usa variables categóricas ignorando todas las numéricas. El modelo final no sirve para predicción real :(
